elastic/elastic-agent
# Add an E2E test to ensure we can install Elastic Defend

Enrolls the agent in Fleet, installs the Elastic Defend package with the Fleet package policies API, then polls the agent to ensure endpoint starts and stays healthy, then uninstalls the agent and confirms the endpoint was uninstalled as well.

To test this locally:

```sh
mage integration:auth
EXTERNAL=true PLATFORMS=linux/amd64,linux/arm64 PACKAGES=targz SNAPSHOT=true mage package
TEST_INTEG_TIMESTAMP=true AGENT_VERSION=8.10.0-SNAPSHOT mage integration:single TestEndpointSecurity
```

- Closes https://github.com/elastic/elastic-agent/issues/2948
- Closes https://github.com/elastic/elastic-agent/issues/3060

This PR also includes a fix for https://github.com/elastic/elastic-agent/issues/3060, which was introduced in https://github.com/elastic/elastic-agent/pull/2889. Uninstalling the agent would fail with an error when endpoint was running with the error `Error: failed to remove installation directory (/Library/Elastic/Agent): unlinkat /Library/Elastic/Agent: directory not empty`.

The problem is the uninstall of endpoint became asynchronous and it would write log at the point where the uninstall command expected the installation directory to be empty. I fixed this by disabling retries when running the `elastic-agent uninstall` command so the execution of the `endpoint-security` uninstall command would be synchronous. I don't think we want infinite retries from shell commands anyway since they can be run by automated tools, better to fail error with the error and let any automation retry as desired.
